doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,615 @@
+  thread-safety.
+- An immutable cursor object (such that `.next()` returns a new cursor
+  object), could increase optimisation complexity for the manager
+  implementation, e.g. managing query cache sharing, lifetime and
+  updates across multiple "live" cursors.
+- Destruction of a cursor object could be used as a signal to the
+  manager that no further pages are required, allowing e.g. cache
+  cleanup.
+
+## Considering batching
+
+Additional considerations arise when considering how batching can be
+handled algongside paging. There are three broad strategies
+
+- Split batching from paging, using a different API method for each.
+- Page through the results from all the batched input elements in
+  lockstep, i.e. whole-batch paging.
+- Page through the results of each batched input element independently,
+  using individual sub-queries within the whole-batch query, i.e.
+  per-element paging.
+
+Each of these batched paging designs combines more or less well with the
+various paging options presented above.
+
+Henceforth the offset-based solution is rejected, given the lack of
+flexibility afforded to the manager's implementation, resulting in poor
+stability and likely manager overhead of record counting.
+
+The above strategies are thus considered in the following sections with
+respect to page token and cursor object designs.
+
+### Batched paging option 1 - Split batching from paging
+
+E.g. Non-paged batch signature
+```python
+# Callback taking index of batch element and all results for that element.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+    batch: List[EntityReference], relationship,
+    successCb: SuccessCallback, errorCb): ...
+```
+
+E.g. Page token paged signature
+
+```python
+def relatedEntities(
+    input: EntityReference, relationship, limit=-1, pageToken=None
+) -> Tuple[AbstractPageToken, List[EntityReferences]]: ...
+```
+
+E.g. Cursor object paged signature
+```python
+class AbstractPage:
+    def get(self) -> List[EntityReferences]:
+        ...
+    def next(self, pageSize):
+        ...
+
+def relatedEntities(
+    input: EntityReference, relationship) -> AbstractPage: ...
+```
+
+#### Pros
+
+- Simpler for manager plugin implementors to reason about (as evidenced
+  by ad-hoc discussions with collaborators).
+- Hosts should usually have a good idea whether batching or paging is
+  more appropriate for the query they are making.
+- Manager and host do not need to consider both paging and batching at
+  the same time, which is a complex convergence of concepts.
+
+#### Cons
+
+- Precludes batching optimisations by the manager when paging is
+  required.
+- Requires implementation of two methods by the manager: one for
+  batched, one for paged.
+- Potential difficulty for some hosts/queries to determine a priori
+  which API (paging or batch) is the most optimal choice.
+- Having the choice adds a foot-gun for host developers to shoot
+  themselves with by accident.
```

Comment:
- Precludes hosts atomically expressing more complex queries that may be required by spreadsheet-type UIs.

Reply:
Added in 86010fc2